src/io/udp_frame.rs

```diff
@@ -0,0 +1,251 @@
+use std::io;
+use std::net::SocketAddr;
+use net::UdpSocket;
+use futures::{Async, Poll, Stream, Sink, StartSend, AsyncSink};
+use futures::sync::BiLock;
+
+/// Encoding of frames via buffers.
+///
+/// This trait is used when constructing an instance of `FramedUdp`. It provides
+/// one type: `Out` for encoding outgoing frames according to a protocol.
+///
+/// Because UDP is a connectionless protocol, the encode method will also be
+/// responsible for determining the remote host to which the datagram should be
+/// sent
+///
+/// The trait itself is implemented on a type that can track state for decoding
+/// or encoding, which is particularly useful for streaming parsers. In many
+/// cases, though, this type will simply be a unit struct (e.g. `struct
+/// HttpCodec`).
+pub trait CodecUdp {
+
+    /// The type of frames to be encoded.
+    type Out;
+    
+    /// The type of decoded frames.
+    type In;
+
+
+    /// Encodes a frame into the buffer provided.
+    ///
+    /// This method will encode `msg` into the byte buffer provided by `buf`.
+    /// The `buf` provided is an internal buffer of the `Framed` instance and
+    /// will be written out when possible. 
+    ///
+    /// The encode method also determines the destination to which the buffer should
+    /// be directed, which will be returned as a SocketAddr;
+    fn encode(&mut self, msg: &Self::Out, buf: &mut Vec<u8>) -> SocketAddr;
+    
+    /// Attempts to decode a frame from the provided buffer of bytes.
+    ///
+    /// This method is called by `FramedUdp` on a single datagram which has been
+    /// read from a socket. 
+    ///
+    /// It is required that the decode method empty the read buffer in every call to
+    /// decode, as the next poll_read that occurs will write the next datagram
+    /// into the buffer, without regard for what is already there. 
+    ///
+    /// If the bytes look valid, but a frame isn't fully available yet, then
+    /// `Ok(None)` is returned. This indicates to the `Framed` instance that
+    /// it needs to read some more bytes before calling this method again.
+    /// In such a case, it is decode's responsibility to copy the data
+    /// into their own internal buffer for future use.
+    ///
+    /// Finally, if the bytes in the buffer are malformed then an error is
+    /// returned indicating why. This informs `Framed` that the stream is now
+    /// corrupt and should be terminated.
+    ///
+    fn decode(&mut self, src: &SocketAddr, buf: &[u8]) -> Result<Option<Self::In>, io::Error>;
```

Comment:
I'm a little worried about this returning an `Option` because I'm not sure what the implementation of `Framed` can do if it returns `None`. Right now it just appends more data to the internal vector but the original `SocketAddr` is lost, meaning that this can be called with data from two different addresses (which I don't think is desired).

I was thinking this would return `Self::In` instead of an `Option` to signify that it always needed to succeed. The codec could perhaps handle partial frames by buffering internally, though?